### PR TITLE
fix(#185): fix bug update extref cb

### PR DESCRIPTION
**Pull request by AliouDIAITE**
Signed-off-by: Aliou DIAITE <aliou.diaite@rte-france.com>
Signed-off-by: SABATIER Philippe Ext <philippe.sabatier@rte-france.com>

### CODE REVIEW

File: sct-commons/src/main/java/org/lfenergy/compas/sct/commons/scl/ied/AbstractLNAdapter.java [hunk 1]
```diff
@@ -574,45 +531,72 @@ public TExtRef checkExtRefInfoCoherence(@NonNull ExtRefInfo extRefInfo) throws S
...
-            return extRef;
-        }
-        List<ControlBlock<?>> cbs = anLNAdapter.getControlSetByExtRefInfo(extRefInfo);
-        isCoherent = !cbs.isEmpty() && cbs.stream()
-                .anyMatch(controlBlock -> controlBlock.getName().equals(sourceInfo.getSrcCBName()));
-        if(!isCoherent){
-            String msg = String.format("Unknown control block %s in %s%s/%s%s%s",
-                    sourceInfo.getSrcCBName(), binderIedName,binderLdInst,
-                    binderLnPrefix , binderLnClass, binderLnInst);
+        List<AbstractLNAdapter<?>> aLNAdapters = new ArrayList<>();
```
Comment (samirromdhani) [inline sct-commons/src/main/java/org/lfenergy/compas/sct/commons/scl/ied/AbstractLNAdapter.java:548]:
```java
        List<AbstractLNAdapter<?>> aLNAdapters = new ArrayList<>();
        if(StringUtils.isBlank(sourceInfo.getSrcLNClass())){
            aLNAdapters.addAll(binderLDeviceAdapter.getLNAdaptersInclundigLN0());
        } else {
            if(TLLN0Enum.LLN_0.value().equals(sourceInfo.getSrcLNClass())){
                aLNAdapters.add(binderLDeviceAdapter.getLN0Adapter());
            } else {
                aLNAdapters.add(binderLDeviceAdapter.getLNAdapter(sourceInfo.getSrcLNClass(), sourceInfo.getSrcLNInst(), sourceInfo.getSrcPrefix()));
            }
        }
```
This part can be extracted from `checkExtRefInfoCoherence` as a function that collecting LNAdapter objects for specific LDeviceAdapter and LN attributes (lnClass, inst and prefix)

**Review by samirromdhani — APPROVED** [review]

File: sct-commons/src/main/java/org/lfenergy/compas/sct/commons/scl/ied/AbstractLNAdapter.java [hunk 2]
```diff
@@ -413,37 +398,45 @@ protected List<ControlBlock<?>> getControlBlocks(List<TDataSet> tDataSets, TServ
...
-        List<? extends TControl> ls = new ArrayList<>();
-        if (TGSEControl.class.equals(cls) && isLN0()) {
-            ls = ((LN0) currentElem).getGSEControl();
-        } else if (TSampledValueControl.class.equals(cls) && isLN0()) {
-            ls = ((LN0) currentElem).getSampledValueControl();
-        } else if (TReportControl.class.equals(cls)) {
-            ls = currentElem.getReportControl();
-        }
-
+    protected <V> List<? extends TControl> lookUpControlBlocksByDataSetRef(@NonNull String dataSetRef, Class<V> cls) {
```
Comment (gleizesDor) [inline sct-commons/src/main/java/org/lfenergy/compas/sct/commons/scl/ied/AbstractLNAdapter.java:404]: I am wondering if this method is really usefull ?

File: sct-commons/src/main/java/org/lfenergy/compas/sct/commons/scl/ied/AbstractLNAdapter.java [hunk 3]
```diff
@@ -360,50 +344,51 @@ public List<ControlBlock<?>> getControlBlocksForMatchingFCDA(@NonNull ExtRefInfo
...
-                );
-            }
+    /**
+     * finds all Control Blocks by dataSetRef and a Service Type (GOOSE, SampleValue, Report) or all Service Types.
+     * @param dataSetRef DatSet value searched
+     * @param serviceType service type to be filtered
+     * @return all Control Blocks matching dataSetRef and a Service Type or all Service Types
+     */
+    private List<ControlBlock<?>> getControlBlocksByDataSetRef(String dataSetRef, TServiceType serviceType) {
+        List<ControlBlock<?>> controlBlocks = new ArrayList<>();
```
Comment (gleizesDor) [inline sct-commons/src/main/java/org/lfenergy/compas/sct/commons/scl/ied/AbstractLNAdapter.java:360]: You can work with streams and use "concat" to concat them together

File: sct-commons/src/main/java/org/lfenergy/compas/sct/commons/scl/ied/AbstractLNAdapter.java (hunk 1 above)
Comment (AliouDIAITE) [reply sct-commons/src/main/java/org/lfenergy/compas/sct/commons/scl/ied/AbstractLNAdapter.java]: ok

**Review by gleizesDor — APPROVED** [review]

File: sct-commons/src/main/java/org/lfenergy/compas/sct/commons/scl/ied/AbstractLNAdapter.java (hunk 3 above)
Comment (AliouDIAITE) [reply sct-commons/src/main/java/org/lfenergy/compas/sct/commons/scl/ied/AbstractLNAdapter.java]: ok

**Pull request merged**